### PR TITLE
[SPARK-6227] [MLlib] [PySpark] Implement PySpark wrappers for SVD and PCA

### PR DESCRIPTION
**Pull request by MechCoder**
Singular Value Decomposition wrappers are missing in PySpark. Since the base for a RowMatrix has been laid writing the wrappers becomes straightforward. Will follow up with the PCA Wrappers in another PR.

### CODE REVIEW

**Comment by MechCoder** [comment]
ping @mengxr

**Comment by MechCoder** [comment]
Actually I'll add the PCA wrappers in this PR as well.

**Comment by MechCoder** [comment]
All right this PR is ready for review.

cc: @dusenberrymw @mengxr

File: docs/mllib-dimensionality-reduction.md [hunk 1]
```diff
@@ -128,6 +128,26 @@ quick-start guide. Be sure to also include *spark-mllib* to your build file as
 a dependency.
 
 </div>
+<div data-lang="python" markdown="1">
+{% highlight python %}
+from pyspark.mllib.linalg import Matrix
```
Comment (dusenberrymw) [inline docs/mllib-dimensionality-reduction.md]: Minor: This import isn't needed.

File: docs/mllib-dimensionality-reduction.md [hunk 2]
```diff
@@ -218,6 +238,30 @@ public class PCA {
...
 
 </div>
+
+<div data-lang="python" markdown="1">
+
+The following code demonstrates how to compute principal components on a `RowMatrix`
+and use them to project the vectors into a low-dimensional space.
+
+{% highlight python %}
+from pyspark.mllib.linalg import Matrix
```
Comment (dusenberrymw) [inline docs/mllib-dimensionality-reduction.md]: Minor: Same as above, this import isn't needed.

File: docs/mllib-dimensionality-reduction.md [hunk 3]
```diff
@@ -218,6 +238,30 @@ public class PCA {
...
+from pyspark.mllib.linalg.distributed import RowMatrix
+from numpy.random import RandomState
+
+# Generate random data with 50 samples and 30 features.
+rng = RandomState(0)
+mat = sc.parallelize(rng.randn(50, 30))
+rm = RowMatrix(mat)
```
Comment (dusenberrymw) [inline docs/mllib-dimensionality-reduction.md]: Minor: I might change `mat` here to `data`, and then change `rm` to `mat` just to be consistent.

File: python/pyspark/mllib/linalg/distributed.py [hunk 4]
```diff
@@ -151,6 +151,112 @@ def numCols(self):
         """
         return self._java_matrix_wrapper.call("numCols")
 
+    def computeSVD(self, k, computeU=False, rCond=1e-9):
+        """
+        Computes the singular value decomposition of the RowMatrix.
+
+        The given row matrix A of dimension (m X n) is decomposed into U * s * V'T where
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Like we talked about in 6485, probably best to trim the docstring lines down to 72 characters. :)

File: python/pyspark/mllib/linalg/distributed.py [hunk 5]
```diff
@@ -151,6 +151,112 @@ def numCols(self):
...
+             eigenvectors of (A X A')
+        * s: DenseVector consisting of square root of the eigenvalues (singular values)
+             in descending order.
+        * v: (n X k) (right singular vectors) is a Matrix whose columns are the
+             eigenvectors of (A' X A)
+
+        For more specific details on implementation, please refer the scala documentation.
+
+        :param k: Set the number of singular values to keep.
+        :param computeU: Whether of not to compute U. If set to be True, then U is computed
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Very minor: Whether _or_ not...

File: python/pyspark/mllib/linalg/distributed.py [hunk 6]
```diff
@@ -151,6 +151,112 @@ def numCols(self):
...
+        >>> rm = RowMatrix(sc.parallelize(data))
+        >>> svd_model = rm.computeSVD(2, True)
+        >>> svd_model.U.rows.collect()
+        [DenseVector([-0.7071, 0.7071]), DenseVector([-0.7071, -0.7071])]
+        >>> svd_model.s
+        DenseVector([3.4641, 3.1623])
+        >>> svd_model.V
+        DenseMatrix(3, 2, [-0.4082, -0.8165, -0.4082, 0.8944, -0.4472, 0.0], 0)
+        """
+        j_model = self._java_matrix_wrapper.call("computeSVD", int(k), computeU, float(rCond))
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Converting `computeU` to a boolean explicitly, `bool(computeU)`, could be useful to allow `0` and `1` to be used.  Alternatively, we could check and enforce a boolean value here, otherwise we get a less-useful error message from the Scala side.

File: python/pyspark/mllib/linalg/distributed.py [hunk 7]
```diff
@@ -151,6 +151,112 @@ def numCols(self):
...
+        Multiplies the given RowMatrix with another matrix.
+
+        :param matrix: Matrix to multiply with.
+        :returns: RowMatrix
+
+        >>> rm = RowMatrix(sc.parallelize([[0, 1], [2, 3]]))
+        >>> rm.multiply(DenseMatrix(2, 2, [0, 2, 1, 3])).rows.collect()
+        [DenseVector([2.0, 3.0]), DenseVector([6.0, 11.0])]
+        """
+        # TODO: Only DenseMatrices are supported on the Scala side.
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Good call, it would definitely be helpful to check that `matrix` is a `DenseMatrix` here.

File: python/pyspark/mllib/linalg/distributed.py [hunk 8]
```diff
@@ -352,6 +458,56 @@ def toBlockMatrix(self, rowsPerBlock=1024, colsPerBlock=1024):
...
+             eigenvectors of (A X A')
+        * s: DenseVector consisting of square root of the eigenvalues (singular values)
+             in descending order.
+        * v: (n X k) (right singular vectors) is a Matrix whose columns are the
+             eigenvectors of (A' X A)
+
+        For more specific details on implementation, please refer the scala documentation.
+
+        :param k: Set the number of singular values to keep.
+        :param computeU: Whether of not to compute U. If set to be True, then U is computed
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Very minor: Whether _or_ not...

File: python/pyspark/mllib/linalg/distributed.py [hunk 9]
```diff
@@ -352,6 +458,56 @@ def toBlockMatrix(self, rowsPerBlock=1024, colsPerBlock=1024):
...
+        >>> svd_model = irm.computeSVD(2, True)
+        >>> svd_model.U.rows.collect() # doctest: +NORMALIZE_WHITESPACE
+        [IndexedRow(0, [-0.707106781187,0.707106781187]),\
+        IndexedRow(1, [-0.707106781187,-0.707106781187])]
+        >>> svd_model.s
+        DenseVector([3.4641, 3.1623])
+        >>> svd_model.V
+        DenseMatrix(3, 2, [-0.4082, -0.8165, -0.4082, 0.8944, -0.4472, 0.0], 0)
+        """
+        j_model = self._java_matrix_wrapper.call("computeSVD", int(k), computeU, float(rCond))
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Same as above with `computeU`.

File: python/pyspark/mllib/linalg/distributed.py [hunk 10]
```diff
@@ -352,6 +458,56 @@ def toBlockMatrix(self, rowsPerBlock=1024, colsPerBlock=1024):
...
+        return SingularValueDecomposition(j_model)
+
+    def multiply(self, matrix):
+        """
+        Multiplies the given IndexedRowMatrix with another matrix.
+
+        :param matrix: Matrix to multiply with.
+        :returns: IndexedRowMatrix
+
+        >>> data = IndexedRowMatrix(sc.parallelize([(0, (0, 1)), (1, (2, 3))]))
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Minor: `data` -> `mat` for consistency.

**Comment by dusenberrymw** [comment]
Great work, @MechCoder!  I left some very small comments, and otherwise it looks good.

File: python/pyspark/mllib/linalg/distributed.py (hunk 7 above)
Comment (MechCoder) [inline python/pyspark/mllib/linalg/distributed.py]: We should either have a _convert_to_matrix function (to support list of lists, and numpy arrays) or raise an error for non-Dense matrices, but I'm kind of inclined towards the former.
(This can of course be worked upon later)

**Comment by MechCoder** [comment]
Thanks for the reviews, I have addressed your comments. Do you have anything else?

File: python/pyspark/mllib/linalg/distributed.py (hunk 7 above)
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: Yeah I agree that a `_convert_to_matrix` function would be useful, but I think that might be better in a separate PR.  For this one, I would just make sure it it is a `DenseMatrix`.

File: python/pyspark/mllib/linalg/distributed.py [hunk 11]
```diff
@@ -352,6 +465,59 @@ def toBlockMatrix(self, rowsPerBlock=1024, colsPerBlock=1024):
...
+        Multiplies the given IndexedRowMatrix with another matrix.
+
+        :param matrix: Matrix to multiply with.
+        :returns: IndexedRowMatrix
+
+        >>> mat = IndexedRowMatrix(sc.parallelize([(0, (0, 1)), (1, (2, 3))]))
+        >>> mat.multiply(DenseMatrix(2, 2, [0, 2, 1, 3])).rows.collect()
+        [IndexedRow(0, [2.0,3.0]), IndexedRow(1, [6.0,11.0])]
+        """
+        return IndexedRowMatrix(self._java_matrix_wrapper.call("multiply", matrix))
```
Comment (dusenberrymw) [inline python/pyspark/mllib/linalg/distributed.py]: I'd check that `matrix` is a `DenseMatrix` here as well.

**Comment by dusenberrymw** [comment]
@MechCoder I'd just add the `DenseMatrix` checks, and then this will be great.  Thanks!

**Comment by cavaunpeu** [comment]
any progress on this @dusenberrymw @MechCoder? it would be really helpful if I could do matrix multiplication in pyspark.

**Comment by MechCoder** [comment]
@cavaunpeu Thanks for the ping! I think I've addressed the pending diff comment.

It will take me some time to refresh the knowledge of the codebase. Can @MLnick or @holdenk give a final pass?

**Comment by MechCoder** [comment]
Bump?

**Comment by MLnick** [comment]
@MechCoder thanks for updating this - may need to wait until after 2.0 release for review.

**Comment by holdenk** [comment]
Now that its past the 2.0 release should we maybe take another look @MLNick / @davies?

File: docs/mllib-dimensionality-reduction.md [hunk 12]
```diff
@@ -84,6 +84,25 @@ quick-start guide. Be sure to also include *spark-mllib* to your build file as
 a dependency.
 
 </div>
+<div data-lang="python" markdown="1">
+{% highlight python %}
+from pyspark.mllib.linalg.distributed import RowMatrix
```
Comment (holdenk) [inline docs/mllib-dimensionality-reduction.md]: Now days we tend write new examples separately and then use the include example syntax to bring them

File: docs/mllib-dimensionality-reduction.md [hunk 13]
```diff
@@ -125,6 +144,29 @@ Refer to the [`RowMatrix` Java docs](api/java/org/apache/spark/mllib/linalg/dist
 {% include_example java/org/apache/spark/examples/mllib/JavaPCAExample.java %}
 
 </div>
+
+<div data-lang="python" markdown="1">
+
+The following code demonstrates how to compute principal components on a `RowMatrix`
+and use them to project the vectors into a low-dimensional space.
+
+{% highlight python %}
```
Comment (holdenk) [inline docs/mllib-dimensionality-reduction.md]: Same as above

File: python/pyspark/mllib/linalg/distributed.py [hunk 14]
```diff
@@ -303,6 +303,121 @@ def tallSkinnyQR(self, computeQ=False):
         R = decomp.call("R")
         return QRDecomposition(Q, R)
 
+    def computeSVD(self, k, computeU=False, rCond=1e-9):
```
Comment (holdenk) [inline python/pyspark/mllib/linalg/distributed.py]: Would be good to add a since annotation here

File: python/pyspark/mllib/linalg/distributed.py [hunk 15]
```diff
@@ -303,6 +303,121 @@ def tallSkinnyQR(self, computeQ=False):
...
+             columns are the eigenvectors of (A X A')
+        * s: DenseVector consisting of square root of the eigenvalues
+             (singular values) in descending order.
+        * v: (n X k) (right singular vectors) is a Matrix whose columns
+             are the eigenvectors of (A' X A)
+
+        For more specific details on implementation, please refer
+        the scala documentation.
+
+        :param k: Set the number of singular values to keep.
```
Comment (holdenk) [inline python/pyspark/mllib/linalg/distributed.py]: It might be good to copy the longer description from RowMatrix for the k param

File: python/pyspark/mllib/linalg/distributed.py [hunk 16]
```diff
@@ -303,6 +303,121 @@ def tallSkinnyQR(self, computeQ=False):
...
+        >>> svd_model.V
+        DenseMatrix(3, 2, [-0.4082, -0.8165, -0.4082, 0.8944, -0.4472, 0.0], 0)
+        """
+        j_model = self._java_matrix_wrapper.call(
+            "computeSVD", int(k), bool(computeU), float(rCond))
+        return SingularValueDecomposition(j_model)
+
+    def computePrincipalComponents(self, k):
+        """
+        Computes the k principal components of the given row matrix
```
Comment (holdenk) [inline python/pyspark/mllib/linalg/distributed.py]: It might be good to copy the warnings form RowMatrix here as well.

File: python/pyspark/mllib/linalg/distributed.py [hunk 17]
```diff
@@ -303,6 +303,121 @@ def tallSkinnyQR(self, computeQ=False):
...
+        """
+        if not isinstance(matrix, DenseMatrix):
+            raise ValueError("Only multiplication with DenseMatrix "
+                             "is supported.")
+        j_model = self._java_matrix_wrapper.call("multiply", matrix)
+        return RowMatrix(j_model)
+
+
+class SingularValueDecomposition(JavaModelWrapper):
+    """Wrapper around the SingularValueDecomposition scala case class"""
```
Comment (holdenk) [inline python/pyspark/mllib/linalg/distributed.py]: Probably add a versionAdded

**Review by holdenk — COMMENTED** [review]
Thanks for working on this and sorry this fell through the cracks post 2.0. I've left some initial comments - likely the same comments apply to the indexed one as well.

**Comment by MechCoder** [comment]
Thanks for the reviews @holdenk . Unfortunately I will not be able to work on this anytime soon. Feel free to cherry-pick the commits, (if you wish)

**Comment by holdenk** [comment]
@MechCoder Thanks! I'll look around and see if anyone else is interested in taking this over and bringing it to the finish line otherwise I'll pick it up myself after OSCON :)

**Comment by HyukjinKwon** [comment]
Ping @MechCoder, are you able to proceed this PR and address the comments above? If not it might be good to close this for now.

**Comment by MLnick** [comment]
Note I revived this at #17621 based on @MechCoder's work.

**Comment by SixAlien3** [comment]
@MLnick Hi, I'm interesting in this PySpark wrapper for SVD. How many columns can this support? Cuz I see in the old document it can only support columns <1000. How about this wrapper?
